x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/topn/TopNOperator.java

```diff
@@ -88,12 +87,53 @@ static final class Row implements Accountable, Releasable {
 
         @Override
         public long ramBytesUsed() {
-            return SHALLOW_SIZE + keys.ramBytesUsed() + orderByCompositeKeyAscending.size() / Byte.SIZE + values.ramBytesUsed();
+            return SHALLOW_SIZE + keys.ramBytesUsed() + bytesOrder.ramBytesUsed() + values.ramBytesUsed();
         }
 
         @Override
         public void close() {
-            Releasables.closeExpectNoException(keys, values);
+            Releasables.closeExpectNoException(keys, values, bytesOrder);
+        }
+    }
+
+    static final class BytesOrder implements Releasable, Accountable {
+        private static final long BASE_RAM_USAGE = RamUsageEstimator.shallowSizeOfInstance(BytesOrder.class);
+        private final CircuitBreaker breaker;
+        final List<SortOrder> sortOrders;
+        final int[] endOffsets;
+
+        BytesOrder(List<SortOrder> sortOrders, CircuitBreaker breaker, String label) {
+            this.breaker = breaker;
+            this.sortOrders = sortOrders;
+            breaker.addEstimateBytesAndMaybeBreak(memoryUsed(sortOrders.size()), label);
+            this.endOffsets = new int[sortOrders.size()];
+        }
+
+        /**
+         * Returns true if the byte at the given position is ordered ascending; otherwise, return false
+         */
+        boolean isByteOrderAscending(int bytePosition) {
+            int index = Arrays.binarySearch(endOffsets, bytePosition);
+            if (index < 0) {
+                index = -1 - index;
+            }
+            return sortOrders.get(index).asc();
+        }
+
+        private long memoryUsed(int numKeys) {
+            return BASE_RAM_USAGE + RamUsageEstimator.alignObjectSize(
+                (long) RamUsageEstimator.NUM_BYTES_ARRAY_HEADER + (long) Integer.BYTES * numKeys
```

Comment:
Maybe a comment would help to explain that we account for `endOffsets` but not for `sortOrders`.